### ISSUE TITLE
False negative when a variable is used within the type expression passed as first argument to `cast` call

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#5429]
**Describe the bug**
Copied from reply in #5391 --

> We can have --
> ```python
> from typing import cast
> 
> type_: type[str] = globals()[""]
> value = cast(type_, None)
> reveal_type(value)  # "str"
> ```
> `str` can be replaced by whatever concrete type we can think of, and revealed type will also change accordingly. Except --
> ```python
> from typing import Any, cast
> 
> type_: type[Any] = globals()[""]
> value = cast(type_, None)
> reveal_type(value)  # "type"???
> ```

The revealed type in the second sample should be `Any` instead of `type`.

**To Reproduce**
Type-check the second sample and watch the revealed type for `value` in the last line.

**Expected behavior**
1. Mark the usage of `cast(...)` as an error if the first argument is not a type, just like what Mypy does.
2. Allow variable whose type is a subtype of `type` as the first argument in `cast(...)`, and let the result type be `Any`/`Unknown` when type of the first argument is `type[Any]`/`type[Unknown]`.

Either way is acceptable.

**Code or Screenshots**
```python
from typing import Any, cast

type_: type[Any] = globals()[""]
# -- or --
# type_: type = globals()[""]
value = cast(type_, None)
reveal_type(value)  # "type"???
```
![image](https://github.com/microsoft/pyright/assets/15389744/d731af21-efe2-4e5d-9b6c-2a2df8760c3d)

**VS Code extension or command-line**
Pylance v2023.6.40
Pyright v1.1.315

**Additional context**
None.

**Comment by cainmagi** [comment]
I can answer your question. This is not a bug because you made a mistake when understanding `Any` and `type`.

First, `type[Any]` is already a typehint. It means "any" type, not the type of `Any`. So your usage is equivalent to:
```python
cast(type[Any], None)
```
If you annotate a variable with `type[Any]`, it should be comprehended as an arbitrary `type`:
```python
# Type[Any] can be used on any type
val: type[Any] = type(1)  # type[int] is compatible with type[Any]
val: type[Any] = type("1")  # type[str] is compatible with type[Any]
```

--------------

Another thing is that, `Any` is not a `type` but an object. so you cannot expect `type[Any]` behave like using `type[...]` with other types:
```python
import inspect
from typing import Any

isinstance(Any, type)  # False
inspect.isclass(Any)  # False
isinstance(int, type)  # True
inspect.isclass(int)  # True
```

So the example you provided should be modified like this:
```python
from typing import Any, cast

type_ = type(Any)
value = cast(type_, None)
reveal_type(value)  # "Any"
```

**Comment by Azureblade3808** [comment]
> So your usage is equivalent to:
> 
> ```python
> cast(type[Any], None)
> ```

Incorrect. You are replacing a typed variable with its type, which should nearly never work.

What is equivalent to `cast(type[Any], None)` needs to be `cast(type(type_), None)` instead of `cast(type_, None)`, presuming that `cast` can accept an expression as its first argument.

**Comment by cainmagi** [comment]
> Incorrect. You are replacing a typed variable with its type, which should nearly never work.

I know `cast(...)` can accept any type expression as the first argument. I only want to let you know that `type[Any]` should be treated as `type` or `type[Unknown]`. They can be used equivalently.

 ```python
from typing import Any


class Customized:
    ...


def func(val: type):
    ...


def func2(val: type[Any]):
    ...


func(Customized)  # Acceptable without errors because `Customized` is an instance of `type`.
func2(Customized)  # Also acceptable without errors because `type[Any]` is the same thing of `type`.
```

I cannot guarantee whether this feature will be preserved after Python3.11, however, at least it is comprehended as above in the current stable version.

**Comment by cainmagi** [comment]
Hello! Finally, I made some interesting experiemnts. Hopefully they can help you:

```python
from typing import Any, cast
from typing_extensions import reveal_type


cast(type[Any], None)
# This is equivalent to the following thing.
type_ = type[Any]
value = cast(type_, None)



class Type_:
    ...


type2_: type[Any] = Type_  # Annotated as type[Type_]
value2 = cast(type2_, None)  # Since type_ is type[Type_], valus is annotated as Type_


type3_ = type
reveal_type(type3_)  # Annotated as type[type]
value3 = cast(type3_, None)  # Annotated as type
# This is reasonable, because the first arg of cast is type3_, not type[type3_].

reveal_type(cast(type[type3_], None))  # Annotated as type[type], equivalent to type3_


type4_: type[Any] = type  # Any is narrowed as type here.
reveal_type(type4_)  # Annotated as type[type]
value4 = cast(type4_, None)  # Annotated as type
# This is reasonable, because type3_ and type4_ are the same thing.


type5_: type[type[Any]] = type  # type[Any] is narrowed as type!
reveal_type(type5_)  # Annotated as type[type]
value5 = cast(type5_, None)  # Annotated as type
# This is werid, because type5_ is narrowed as the same thing of type3_.


type6_: type[type[type[Any]]] = type  # Even more werid, type[type[Any]] is narrowed as type.
reveal_type(type6_)  # Annotated as type[type]
value6 = cast(type6_, None)  # Annotated as type
# This is werid, because type6_ is narrowed as the same thing of type3_.


type7_: type[Any] = globals()[""]  # Compatible with an arbitrary class.
reveal_type(type7_)  # Annotated as type, as expected
value7 = cast(type7_, None)  # Should not be type
value7_ = cast(type[Any], None)  # Should not be type, Any is incorrecly narrowed as type.
value7__ = cast(type, None)  # Should not be type, Unknown is incorrecly narrowed as type.
# value7 and value7_ should be equivalent to this:
def func(val: Any):
    value7___ = cast(type(val), None)  # Annotated as Any, as expected.
    
    # Further checks.
    reveal_type(type(val))  # Annotated as Any, not type[Any]!
    reveal_type(type(object()))  # Annotated as type[object].
```

### My understanding

Now I also admit that this behavior is very weird. Because `type(val: Any)` is treated as `Any`, but `type(val: object)` is treated as `type[object]`.

I think the reason causing this behavior maybe the following thing:

```python
type is type(type)   # True
```

`type` is a special `type` instance which type is itself. This fact causes that you can treat `type` as an instance of `type` or a class named `type`. Therefore, both `Any` and `type[Any]` can be narrowed as `type`, which may even cause a recursive behavior.

In your example, `type_: type[Any]` is treated as an arbitrary instance of `type`, when it is fed to `cast`, its returned value should be annotated as `Any`. However, because `type` can be an instance of the class `type`, the linter chooses `type` as the inferred result.

I agree that this issue should be fixed.

**Comment by Azureblade3808** [comment]
As a reference, the behavior of `type(val: Any)` resulting in `Any` has been discussed in microsoft/pylance-release#747 and Eric's explanation is [here](https://github.com/microsoft/pylance-release/issues/747#issuecomment-747181804).

**Comment by cainmagi** [comment]
> As a reference, the behavior of `type(val: Any)` resulting in `Any` has been discussed in [microsoft/pylance-release#747](https://github.com/microsoft/pylance-release/issues/747) and Eric's explanation is [here](https://github.com/microsoft/pylance-release/issues/747#issuecomment-747181804).

Understood. I would like to add a further inspection:

```python
from typing_extensions import reveal_type


def func(type_: type, *args, **kwargs):
    val = type_(*args, **kwargs)
    reveal_type(val)  # Any, as expected
```
If a function variable is annotated as `type` or `type[Any]`, Its corresponding instance can be correctly annotated as `Any`. This behavior is different from using `cast(type, ...)`, which is annotated as `type`. At least I think `cast(type, ...)` should have the same behavior as the instantiation of a variable annotated as `type`.

**Comment by erictraut** [comment]
@cainmagi, in your code sample above, you're attempting to using variables within a type expression, which is not allowed. Mypy correctly flags these as errors. Pyright was a bit too permissive in this case. I'll fix this so pyright matches mypy's behavior.

**Comment by erictraut** [comment]
I've changed pyright's logic for the `cast` call to properly enforce that the first argument must be a valid type expression. This means that variables are no longer allowed, just as they are not within a type annotation.

Pyright treats `type[Any]` as equivalent to `type`, so if you cast an expression to `type[Any]`, the resulting type will be `type[Any]` (which may be printed simply as `type`). So that part is working correctly.

**Comment by cainmagi** [comment]
> @cainmagi, in your code sample above, you're attempting to using variables within a type expression, which is not allowed. Mypy correctly flags these as errors. Pyright was a bit too permissive in this case. I'll fix this so pyright matches mypy's behavior.

In the previous versions, the type checker has already enforced such requirements for `tp` in `cast(tp, ...)`. I think the reason why the checker does not raise errors may be because the variable is detected to be a `type` instance.

Thank you for emphasizing this rule. I will test it when the new version is released. If I have any related questions, I will let you know in this thread.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. It will also be included in a future release of pylance.
